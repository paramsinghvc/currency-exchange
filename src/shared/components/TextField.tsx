import React, { FC } from "react";
import styled from "@emotion/styled";
import theme from "shared/theme";

export type sizeType = "small" | "medium" | "large" | "x-large";

export type ITFProps = ITextFieldProps;

export interface ITextFieldProps extends React.HTMLProps<HTMLInputElement | HTMLTextAreaElement> {
  type?: string;
  id?: string;
  value: string | number;
  multiplelines?: boolean;
  onChange: (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void;
  placeholder?: string;
  fullWidth?: boolean;
  error?: string | boolean;
  leftAdornment?: JSX.Element | string;
  rightAdornment?: JSX.Element | string;
  width?: number;
  paddingsize?: sizeType;
  className?: string;
}

const sizeMap = new Map<sizeType, string>([
  ["small", "5px 8px"],
  ["medium", "10px 10px"],
  ["large", "15px 15px"],
  ["x-large", "20px 15px"]
]);

const TextField = styled.div`
  display: inline-flex;
  font-size: 16px;
  border: 2px solid ${theme.borderColor};
  border-radius: ${theme.borderRadius};
  &:focus-within {
    border: 2px solid ${theme.primaryColor};
  }
`;

const InputBox = styled.input<Pick<ITextFieldProps, "paddingsize">>`
  background: none;
  padding: ${props => sizeMap.get(props.paddingsize!)};
  border: none;
  outline: none;
`;

const TextArea = styled.textarea<Pick<ITextFieldProps, "paddingsize">>`
  ${InputBox}
`;

const Adornment = styled.div<Pick<ITextFieldProps, "paddingsize">>`
  padding: ${props => sizeMap.get(props.paddingsize!)};
  &.left {
    border-right: 2px solid ${theme.borderColor};
  }
  &.right {
    border-left: 2px solid ${theme.borderColor};
  }
`;
const TextFieldComp: FC<ITFProps> = props => (
  <TextField {...(props as React.HTMLProps<HTMLDivElement>)}>
    {props.leftAdornment && (
      <Adornment className="left" paddingsize={props.paddingsize}>
        {props.leftAdornment}
      </Adornment>
    )}
    {props.multiplelines ? (
      <TextArea value={props.value} {...(props as React.HTMLProps<HTMLTextAreaElement>)} />
    ) : (
      <InputBox type={props.type} value={props.value} {...(props as React.HTMLProps<HTMLInputElement>)} />
    )}
    {props.rightAdornment && (
      <div className="right" css={Adornment}>
        {props.rightAdornment}
      </div>
    )}
  </TextField>
);

const defaultProps = {
  paddingsize: "large" as sizeType,
  value: "",
  onChange: () => {},
  type: "text"
};

TextFieldComp.defaultProps = defaultProps;

export default TextFieldComp;
