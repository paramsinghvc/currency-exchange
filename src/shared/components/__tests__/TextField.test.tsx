import React from "react";
import TextField from "../TextField";
import renderer from "react-test-renderer";

describe("Should render Button component properly", () => {
  it("should render a text field", () => {
    const textField = renderer.create(<TextField value={"City of stars"} onChange={() => {}} />);
    expect(textField.toJSON()).toMatchSnapshot();
  });

  it("should render a full width text field", () => {
    const textField = renderer.create(<TextField value={"City of stars"} onChange={() => {}} fullWidth />);
    expect(textField.toJSON()).toMatchSnapshot();
  });

  it("should render a text field with adornment", () => {
    const textField = renderer.create(
      <TextField
        value={"City of stars"}
        onChange={() => {}}
        leftAdornment={<span>$</span>}
        rightAdornment={<span>Kg</span>}
      />
    );
    expect(textField.toJSON()).toMatchSnapshot();
  });
});
