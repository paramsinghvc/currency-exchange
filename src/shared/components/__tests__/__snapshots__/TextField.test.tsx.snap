// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Button component properly should render a full width text field 1`] = `
<div
  className="css-1l5b2qj"
  onChange={[Function]}
  type="text"
  value="City of stars"
>
  <input
    className="css-1a7xov3"
    onChange={[Function]}
    type="text"
    value="City of stars"
  />
</div>
`;

exports[`Should render Button component properly should render a text field 1`] = `
<div
  className="css-1l5b2qj"
  onChange={[Function]}
  type="text"
  value="City of stars"
>
  <input
    className="css-1a7xov3"
    onChange={[Function]}
    type="text"
    value="City of stars"
  />
</div>
`;

exports[`Should render Button component properly should render a text field with adornment 1`] = `
<div
  className="css-1l5b2qj"
  onChange={[Function]}
  type="text"
  value="City of stars"
>
  <div
    className="left css-100kda6"
  >
    <span>
      $
    </span>
  </div>
  <input
    className="css-1a7xov3"
    onChange={[Function]}
    type="text"
    value="City of stars"
  />
  <div
    className="right"
    css={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "__emotion_base": "div",
        "__emotion_forwardProp": undefined,
        "__emotion_real": [Circular],
        "__emotion_styles": Array [
          "
  padding: ",
          [Function],
          ";
  &.left {
    border-right: 2px solid ",
          "#F2F2F2",
          ";
  }
  &.right {
    border-left: 2px solid ",
          "#F2F2F2",
          ";
  }
",
        ],
        "defaultProps": undefined,
        "displayName": "Styled(div)",
        "render": [Function],
        "withComponent": [Function],
      }
    }
  >
    <span>
      Kg
    </span>
  </div>
</div>
`;
